Propagate HasSynced properly

lavalamp opened the pull request:
Fix #113763 by:
* Fix informers to track HasSynced properly
* propagate HasSynced to each Handle individually
* Add AsyncTracker to help multi-worker controllers track whether they've synced or not


/kind bug

#### What this PR does / why we need it:

Plumbs the fact of whether an item was in the initial list or not to controllers, where they can choose to track whether they have processed it or not.

```release-note
Shared informers now correctly propagate whether they are synced or not. Individual informer handlers may now check if they are synced or not (new HasSynced method). Library support is added to assist controllers in tracking whether their own work is completed for items in the initial list (AsyncTracker).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

lavalamp committed:
Enable propagration of HasSynced

* Add tracker types and tests
* Modify ResourceEventHandler interface's OnAdd member
* Add additional ResourceEventHandlerDetailedFuncs struct
* Fix SharedInformer to let users track HasSynced for their handlers
* Fix in-tree controllers which weren't computing HasSynced correctly
* Deprecate the cache.Pop function

howardjohn commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -221,15 +224,15 @@ type ResourceEventHandler interface {
 // modifying the objects.
 type ResourceEventHandlerFuncs struct {
-	AddFunc    func(obj interface{})
+	AddFunc    func(obj interface{}, isInInitialList bool)
```
This seems like it would require changes across the ecosystem. would it make sense to add another callback? That should be backwards compat for all cases except `ResourceEventHandlerFuncs{a,b,c}` which is typically a linter error anyways

lavalamp replied:
Yeah, the code can test for which version is passed in.

lavalamp replied:
That would make the tests less scary, for sure ;)

liggitt replied:
yeah... it would be really great to not require a signature from all consumers

lavalamp replied:
OK I think this will accomplish that, the intermediate state was good to prove to me that that I got everything in this package passing this through correctly.

Still needs comment updates and tests.

And it needs to be used by one controller to prove / demonstrate how it works, what controller would you like me to update?

lavalamp commented:
OK shockingly this is now a local change that doesn't break the world!

I'd like feedback on the names before I fix all the comments?

howardjohn commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -526,6 +530,7 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 
 			f.cond.Wait()
 		}
+		isInInitialList := !f.hasSynced_locked()
```
My understanding is `isInInitialList` means "this is from the initial list".

The issue I have as a consumer is that I still can't tell when the initial list is _done_. I can't wait for an Add with `isInInitialList=false`, because there may never be one.

I would have expected this to be `isLastItemInList := f.hasSynced_locked()` (at L539, not here). Then a consumer can handle in their add controller `if isLastItemInList { do what I want on sync }`.

This may also lean towards having an `OnInitiallListCompleteFunc()` callback rather than a bool in the `AddFunc`, but that is stylistic - I don't have a strong preference

lavalamp replied:
What you describe won't work because eventually people start doing work async, knowing the last one isn't helpful, you need to know every one that's still in flight.

I will show how to use this in at least one controller.

The final check is going to look like `informer.HasSynced() && !initialListItems.InFlight()`, the latter part isn't in this PR yet.

howardjohn replied:
My thinking was that each layer needs to propagate some notion of "done". Anytime you add async, you are breaking that, and need to add it back at that layer. The goal of this is to propogate it at the informer handler layer. If you have _another_ layer on top of that (multiple worker queues, in k8s typically), then you will need some higher level solution on top of that which can be done by just knowing once all the event handlers have been called. _how_ exactly to do that depends on the code on top of the informer.

Its a bit hard to discuss the final check since I may be misunderstanding the intended flow, but I am guessing it is keeping track of all the items that come in with isInInitialList=true and then removing them as the queue completes?
 
If so, IMO it would still be preferable to flip the logic. 

If users are using multi worker queues, they _can_ still do this by just keeping track of items that come in before OnInitiallListCompleteFunc triggers. But for users who don't have this setup, they shouldn't have to pay the complexity of tracking items when a simple callback will do IMO.

lavalamp replied:
Users that aren't doing anything async already don't have any problems, so I don't understand why I shouldn't optimize it for people that are doing async things?

lavalamp replied:
(BTW I agree having a separate function to call on completion of the initial list would be better if we were starting from scratch, but I think what I've done here is going to have more minimal changes for consumers)

howardjohn replied:
This might just be a terminology thing - but I am not sure my understanding aligns with yours on non-async being ok as is.

When I am referring to non-async handlers, I mean something like this: https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go#L100, while an async handler is typically `queue.Add`.

With that definition, non-sync users are still broken - they still cannot tell when they are fully synced because the informer internals are also async and don't expose a way to know when the initial list is complete (leading to https://github.com/kubernetes/kubernetes/issues/113761).

> more minimal changes for consumers

My initial thought was just adding a field to `ResourceEventHandlerFuncs`; adding a field seems like its backwards compatible in almost all cases except the cases of keyless struct initialization, which seems rare. But I didn't dig into it much, so maybe there is more issues?

---

To be clear - I think this works and I am far happier to have this than nothing, consider these nitpicks :slightly_smiling_face:

lavalamp replied:
.... that code does N*log(N) work on start up for N webhooks, that's really bad, wow.

It appears to be wrong because HasSynced is called on the shared part of the informer, but not all notifications have made it through to the non-shared stub? That seems like a separate problem that I can also fix.

lavalamp replied:
> this controller doesn't care about the underlying store since it maintains its own index on top of things

Specifically the problem is the lag -- shared informer pipes everything through a channel, HasSynced is correct except for the stuff sitting in that channel. If this were using a non-shared informer it'd actually be correct; in that case HasSynced won't return true until after your Add handler for the last initial item has returned.

howardjohn replied:
That doesn't align with my experience. Both shared and unshared informers suffer the same issue. The issue comes down to `processorListener` which is used by all informers; `processorListener.add()` is async. If it wasn't, then everything would work fine actually (that is, `HasSynced` would also mean "event handlers have seen initial state", not just the store). This is why "Option 1: Make processorListener synchronous" in https://github.com/kubernetes/kubernetes/issues/113763 resolves the issue (although its probably a bad idea)

Here is a test of both shared and unshared that can reproduce it. Note its flaky - fails about 1% of the timer on my machine, so you need to run it with `stress` or similar.

<details>

```go
package main

import (
	"fmt"
	"sync/atomic"
	"testing"
	"time"

	v1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/util/wait"
	"k8s.io/client-go/informers"
	informersv1 "k8s.io/client-go/informers/core/v1"
	"k8s.io/client-go/kubernetes/fake"
	"k8s.io/client-go/tools/cache"
)

func TestInformerUnshared(t *testing.T) {
	kc := fake.NewSimpleClientset(pods...)
	handled := &atomic.Int32{}
	podInformer := informersv1.NewFilteredPodInformer(kc, "", 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, nil)
	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj any) {
			handled.Add(1)
		},
	})

	stop := make(chan struct{})
	go podInformer.Run(stop)
	WaitForCacheSync(stop, podInformer.HasSynced)

	if h := handled.Load(); h != int32(len(pods)) {
		t.Fatalf("got %+v events, expected %v", h, len(pods))
	}
}

func TestInformerShared(t *testing.T) {
	kc := fake.NewSimpleClientset(pods...)
	informers := informers.NewSharedInformerFactory(kc, 0)
	handled := &atomic.Int32{}
	podInformer := informers.Core().V1().Pods().Informer()
	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj any) {
			handled.Add(1)
		},
	})

	stop := make(chan struct{})
	informers.Start(stop)
	WaitForInformerFactorySync(stop, informers)
	WaitForCacheSync(stop, podInformer.HasSynced)

	if h := handled.Load(); h != int32(len(pods)) {
		t.Fatalf("got %+v events, expected %v", h, len(pods))
	}
}

func WaitForInformerFactorySync(stop <-chan struct{}, informerFactory informers.SharedInformerFactory) {
	returnImmediately := make(chan struct{})
	close(returnImmediately)
	_ = wait.PollImmediate(time.Microsecond*10, wait.ForeverTestTimeout, func() (bool, error) {
		select {
		case <-stop:
			return false, fmt.Errorf("channel closed")
		default:
		}
		for _, synced := range informerFactory.WaitForCacheSync(returnImmediately) {
			if !synced {
				return false, nil
			}
		}
		return true, nil
	})
}

// Like standard WaitForCacheSync but more aggressive, to more easily detect race conditions.
func WaitForCacheSync(stop <-chan struct{}, cacheSyncs ...cache.InformerSynced) bool {
	max := time.Millisecond * 100
	delay := time.Millisecond
	f := func() bool {
		for _, syncFunc := range cacheSyncs {
			if !syncFunc() {
				return false
			}
		}
		return true
	}
	for {
		select {
		case <-stop:
			return false
		default:
		}
		res := f()
		if res {
			return true
		}
		delay *= 2
		if delay > max {
			delay = max
		}
		time.Sleep(delay)
	}
}

var pods = []runtime.Object{
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "a",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "b",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "c",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "d",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
}
```

</details>

lavalamp replied:
NewFilteredPodInformer still makes a shared informer: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/informers/core/v1/pod.go#L58-L59

howardjohn replied:
Ahhh... thank you. That explains a lot of confusion I had even beyond this issue. Thanks!

So the "Option 1: Make processorListener synchronous" is actually not _that_ crazy - that is exactly how the non-shared variant works anyways. It still may not be ideal since now unrelated handlers can block each other though (not a concern for my personal use cases, but I can see how that would be for some users).

lavalamp replied:
Well, we can't let one bad handler block others, so that's not feasible. I'll figure out a fix here in a bit.

lavalamp replied:
OK, check out what that controller looks like now.

🙏 that tests pass still ;)

(I need to add some tests)

lavalamp commented:
I'm gonna be out next week, so this will be on hold for a bit unless I feel especially inspired

lavalamp commented:
(I was feeling inspired, all this needs now is more tests and a use of the AsyncTracker. And to be squashed.)

howardjohn reviewed (commented):
Tested this out with some of the rests from the original issues, passes with `handler.HasSync` checked :+1: :+1:

lavalamp commented:
Great news, then! We can probably get this ready to go next week when I'm back.

eahydra commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -559,7 +574,7 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
 	}
 
-	listener := newProcessListener(handler, resyncPeriod, determineResyncPeriod(resyncPeriod, s.resyncCheckPeriod), s.clock.Now(), initialBufferSize)
+	listener := newProcessListener(handler, resyncPeriod, determineResyncPeriod(resyncPeriod, s.resyncCheckPeriod), s.clock.Now(), initialBufferSize, s.HasSynced)
```
hi @lavalamp , 
If the EventHandler is registered before the Informer data synchronization, then this PR can help determine whether the EventHandler is synchronized (processorListener.HasSync return true). But there is another scenario is to register the EventHandler after the Informer data synchronization is completed, and it is also expected to determine whether the EventHandler is synchronized. But this PR doesn't seem to support this scenario. Will this scenario be supported?

https://github.com/kubernetes/kubernetes/blob/fdd7b4ce5fde399d2a3878ca4f5d569e42dff017/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L577-L595

lavalamp replied:
Yeah this is one of the cases I want to fix/test.

lavalamp replied:
This is fixed in the latest push

eahydra replied:
Thank you @lavalamp .
Not sure if the implementation of this place will have a race condition. For example, after the Cache data synchronization is completed, the EventHandler is registered. At this time, it traverses the data in the Cache, initiates an addNotification and calls syncTracker.Start, then syncTracker.count becomes to 1. When the addNotification is consumed, execute syncTracker.Finish, then syncTrakcer.count becomes to 0, just at this time an external goroutine called syncTracker.HasSynced, because upstream hasSynced, and syncTracker.count = 0, so syncTracker.HasSynced returns true.

Maybe can consider passing the total number of objects in the Cache directly to syncTracker.count before initiating addNotification.

WDYT?

lavalamp replied:
We hold the lock on the informer while adding these, which means no one can call the informer's HasSynced method untill we have added them all, which also means the handler's HasSynced method can't be called, since it calls the informer's HasSynced method. I can expand the comment here.

However now that you mention it I need to fix up the tracker to check things in the other order, I think.

lavalamp replied:
Actually if you read the comment it already mostly says this, but I will add your scenario in as another thing that's protected against this way.

eahydra replied:
Yes... I overlooked that..😅

lavalamp commented:
Ready for a review. still needs a squash.

/assign @liggitt 
/assign @deads2k

fedebongio commented:
/triage accepted

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/internal/generic/controller_test.go:
```diff
@@ -406,7 +405,12 @@ func TestInformerList(t *testing.T) {
-	require.True(t, cache.WaitForCacheSync(testContext.Done(), myController.HasSynced))
+	defer func() {
+		testCancel()
+		wg.Wait()
```
@alexzielenski can you review the changes to this test? It appears to have been slightly broken before.

lavalamp replied:
...I made more extensive changes and left some TODOs for you (sorry!)

alexzielenski replied:
test/cel changes + todos lgtm

lavalamp commented:
Someone please review this so I stop fixing more controllers!!

eahydra commented:
I think this PR can really help solve many problems.  I wonder if it will be updated to the old version, such as v1.22?

stevekuznetsov commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:
```diff
@@ -480,6 +480,17 @@ func TestDeltaFIFO_UpdateResyncRace(t *testing.T) {
 	}
 }
 
+func pop2[T any](queue Queue) (T, bool) {
```
This only pops one value - what's the meaning of `2`?

lavalamp replied:
In contrast to the `Pop` function elsewhere, this returns both values.

lavalamp replied:
I added a comment

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -247,6 +301,20 @@ func (r ResourceEventHandlerFuncs) OnDelete(obj interface{}) {
+		detailed.OnAdd(obj, isInInitialList)
+		return
+	}
+	panic("Need at least one implementation of Add")
```
not that I can think of without requiring changes at every use :/

stevekuznetsov replied:
Gotcha, no worries.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -526,6 +530,7 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 
 			f.cond.Wait()
 		}
+		isInInitialList := !f.hasSynced_locked()
```
Things can't get demoted/preempted in the queue, so I think this isn't a problem?

A future change could get merged with a change from the initial list, if you don't want that you have to use the regular fifo. But most uses do want that, since it's more efficient.

stevekuznetsov replied:
Why is the queue important? My read was that `f.populated` was global? I'm considering the case where:

1. `*Reflector.syncWith()` calls `*DeltaFIFO.Replace()` for all items in the original list
1. `*Reflector.ListAndWatch()` begins the watch
1. `*controller.processLoop()` begins calling `*DeltaFIFO.Pop()` to process them
1. `*Reflector.watchHandler()` calls `*DeltaFIFO.Add()` with some new object it saw in the watch stream
1. `*controller.processLoop()` finishes using `*DeltaFIFO.Pop()` to consume all objects from the initial list

The delta FIFO `HasSynced()` would return `true` after 4, even though the initial list had not been processed yet, right? Since `f.populated` gets latched once `Add()` is called. The controller relinquishes the FIFO's lock between each `Pop()`, so the `Add()` could be interleaved?

lavalamp replied:
`populated` is already set to true at the end of Replace()? It's initialPopulationCount going to zero that determines when it is synced.

stevekuznetsov replied:
Ah, jeez, sorry - I took the comment on `HasSynced` at face value and didn't realize `Replace()` also set it. Sorry for the noise!

lavalamp replied:
no worries :)

deads2k commented on pkg/controller/certificates/rootcacertpublisher/publisher.go:
```diff
@@ -58,18 +58,18 @@ func NewPublisher(cmInformer coreinformers.ConfigMapInformer, nsInformer coreinf
 	})
 	e.cmLister = cmInformer.Lister()
-	e.cmListerSynced = cmInformer.Informer().HasSynced
+	e.cmListerSynced = cmHandler.HasSynced
```
This has changed from "is my lister synced" to "is my lister synced and my event handler been synced"? Why change the precondition for starting on this controller to include event handler syncing?

lavalamp replied:
I may have gotten a little overeager on a few of the controllers... I already backed out some. It turns out we just don't actually wait for them to be synced as much as I thought we did

deads2k commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -210,16 +210,70 @@ func (c *controller) processLoop() {
+// events. Check this assumption.
+type ResourceEventHandlerDetailed interface {
+	ResourceEventHandler
+	OnAdd(obj interface{}, isInInitialList bool)
```
Will there be a followup just breaking the contract and using  this?  I have a personal preference for only have the API I want to support.

lavalamp replied:
I added a commit adding notices to the things that would change -- I'm not totally sure it's worth the complaints we'll get but I don't care too much

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go:
```diff
@@ -59,12 +58,17 @@ func NewMutatingWebhookConfigurationManager(f informers.SharedInformerFactory) g
 
+	// Since our processing is synchronous, this is all we need to do to
+	// see if we have processed everything or not.
+	manager.hasSynced = handler.HasSynced
```
This one is worthwhile to keep for correctness.

deads2k commented on staging/src/k8s.io/client-go/tools/cache/synctrack/synctrack.go:
```diff
@@ -0,0 +1,107 @@
+
+// Finished should be called when finished processing a key which was part of
+// the initial list.
+func (t *SingleFileTracker) Finished() {
```
if I've read the code right, this should never go negative and if it does go negative, it means that our internal cache is in an unpredictable state.  If I've read correctly, a utilruntime.handlepanic seems reasonable to trigger from this point.

lavalamp replied:
Yeah, for the single file tracker that's a usage error.

For the async tracker, it's actually OK to just unconditionally Finish() keys. I can at least expand the comments.

lavalamp replied:
I'm not sure I can bring myself to add a panic and I don't want this package to import utilruntime just for that, but comments are improved now.

deads2k replied:
> I'm not sure I can bring myself to add a panic

How will the library author know if they've made an egregious error that reviewers didn't catch without a mechanism to know if these went negative?

lavalamp replied:
Added a panic in latest commit

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/internal/generic/controller.go:
```diff
@@ -188,7 +194,11 @@ func (c *controller[T]) Run(ctx context.Context) error {
+	if hasSynced == nil {
+		return false
+	}
+	return hasSynced.(func() bool)()
```
this is an example of where the handler is queueing objects for async processing we'd need to track actually completed handling the initial set, right?

lavalamp replied:
Yes, I missed this -- fixed in last commit

lavalamp commented:
Rebased, removed some controllers that don't actually need this treatment, I think the remaining ones are OK although I'm not sure I understand why the graph_populater is waiting for the cache sync in a goroutine.

I'm not sure we should make everyone change so the last commit with the deprecation notices is still speculative.

lavalamp commented on pkg/controller/replication/conversion.go:
```diff
@@ -125,13 +125,13 @@ type conversionEventHandler struct {
 }
 
-func (h conversionEventHandler) OnAdd(obj interface{}) {
+func (h conversionEventHandler) OnAdd(obj interface{}, isInInitialList bool) {
```
@liggitt I think this was the only middle layer I had to fix -- IMO this fix is a lot more convenient than adding a second method and making both work

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/configuration/validating_webhook_manager.go:
```diff
@@ -73,29 +76,9 @@ func (v *validatingWebhookConfigurationManager) Webhooks() []webhook.WebhookAcce
-		// Setting initialConfigurationSynced to true, so subsequent checks
-		// would be able to take the fast path on the atomic boolean in a
-		// cluster without any admission webhooks configured.
-		v.initialConfigurationSynced.Store(true)
```
if we're dropping this, should probably drop the initialConfigurationSynced field as well

lavalamp replied:
Pushed a change (I am TRYING my best to not get rid of the N^2 startup behavior in this PR, it's harder to resist every time I look at this file).

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -210,16 +210,82 @@ func (c *controller) processLoop() {
+// ResourceEventHandlerInspecific or ResourceEventHandlerDetailed (preferred).
+//
+// REQUIRED ACTION NOTICE: `OnAdd(obj interface{}, isInInitialList bool)` will
+// be added to this interface in v1.28.
```
If we're going to change the interface, I don't see why we would wait a release and make this interface less usable in the meantime. If changing this interface now has a lot of ripples and is painful to do in this PR, that might be a sign that we should ease that transition, rather than simply delay it.

If we're going to change the signature of OnAdd, I think it would be better to:
1. change the interface now unconditionally
2. leave the member fields in ResourceEventHandlerFuncs as-is (so ResourceEventHandlerFuncs#OnAdd would just drop the isInInitialList param)
3. define ResourceEventHandlerDetailedFuncs type as you did in this PR, so people who want the detailed information can get it

Trying out that approach only required changing one additional OnAdd implementation (in test/integration/client/exec_test.go), because almost all callers use ResourceEventHandlerFuncs

It also dropped several runtime error paths where something that didn't implement an acceptable OnAdd function could get passed around.

lavalamp replied:
OK I pushed this as a new commit. I think it risks more clients having to do something, but it is shorter

> It also dropped several runtime error paths

I think they are replaced by much subtler bugs of failing to propagate isInInitialList, if any middleware exists out there. Bugs would have been caught at initialization time.

lavalamp commented:
I think this is ready to go (but it should be squashed the rest of the way)

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/internal/generic/controller.go:
```diff
@@ -69,12 +76,20 @@ func NewController[T runtime.Object](
 		queue:      nil,
 	}
+	c.hasProcessed.UpstreamHasSynced = func() bool {
+		f := c.notificationsDelivered.Load()
```
fixes one of our issues to take CEL for beta for me, thanks!

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -194,7 +194,7 @@ func (c *controller) processLoop() {
 }
 
-// ResourceEventHandler can handle notifications for events that
+// ResourceEventHandlerUniversal can handle notifications for events that
```
It was, fixed now, thanks!

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller_reconcile.go:
```diff
@@ -128,7 +128,7 @@ func (c *celAdmissionController) reconcilePolicyDefinition(namespace, name strin
 			paramsGVR.Resource,
 			corev1.NamespaceAll,
-			30*time.Second,
+			30*time.Second, // TODO: do we really need to ever resync these?
```
agreed we shouldn't be resyncing these... @alexzielenski, can you make sure this gets added to the list of things to resolve with param instance watching while in alpha?

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:
```diff
@@ -165,6 +166,7 @@ func NewAdmissionController(
 }
 
 func (c *celAdmissionController) Run(stopCh <-chan struct{}) {
+	// TODO: Doesn't this comparison need a lock?
```
yeah, this appears to be supporting stopping and then restarting the admission controller, or tolerating multiple calls to Run(), neither of which are things I would expect.

@alexzielenski, can you add this to the list of bits to revisit and maybe remove?

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -575,27 +590,35 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
 	handle := s.processor.addListener(listener)
 	for _, item := range s.indexer.List() {
-		listener.add(addNotification{newObj: item})
+		// Note that we deliver these notifications with the lock held
+		// and before returning the handle. That means there is never a
```
> Note that we deliver these notifications with the lock held and before returning the handle

nit: I misunderstood this to mean the notifications were delivered *to the handler* before returning, and went off on a rabbit trail because I was pretty sure that wasn't true (spoiler: it's not, there's two async levels before the notifications get to the handler in pop() and run())

maybe rephrase to something like "we enqueue these notifications to the syncTracker with the lock held and before returning the handle" to save future us from rabbit trails

lavalamp replied:
Yeah sorry, "enqueue" is much less confusing, I'll change it

lavalamp replied:
fixed, sorry

liggitt commented on staging/src/k8s.io/client-go/tools/cache/fifo.go:
```diff
@@ -84,8 +84,9 @@ type Queue interface {
-	queue.Pop(func(obj interface{}) error {
+	queue.Pop(func(obj interface{}, isInInitialList bool) error {
 		result = obj
+		// TODO: handle isInInitialList?
```
nit: either clarify what type of handling this is asking for or drop the todo

lavalamp replied:
Fixed (and marked this function for deprecation)

liggitt reviewed (commented):
I'll defer to @deads2k on the approach of propagating the inInitialList bool to the add handler, but if we're going to do that, this looks like a reasonable way to accomplish it. A few optional nits, but this looks correct to me.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -220,15 +220,23 @@ type ResourceEventHandler interface {
+		r.AddDetailedFunc(obj, isInInitialList)
+	} else if r.AddFunc != nil {
 		r.AddFunc(obj)
 	}
```
Ah, is that what you were trying to ask for originally? fixed

lavalamp commented:
Whoops, maybe it actually compiles with this push 😳

lavalamp commented:
/retest

lavalamp commented:
This is squashed. @deads2k?

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -876,11 +901,18 @@ type processorListener struct {
 
-func newProcessListener(handler ResourceEventHandler, requestedResyncPeriod, resyncPeriod time.Duration, now time.Time, bufferSize int) *processorListener {
+// HasSynced returns true if the source informer has synced, and all
+// corresponding events have been delivered.
```
```suggestion
// corresponding events have been delivered to every event handler.
```

is that edit correct?

deads2k replied:
oh, this is the lower level one.  nm.

deads2k commented:
The interface lgtm now, the logic looks fine.  I'm concerned about how we'll know if we've done something incorrect with the sync tracker without any error and without a panic.  This code is used to ensure to synchronization of resources used for server safety and as it stands now, if we had a significant bug, we could improperly report "synchronized==true" without any reporting or failure.  I think that panicking is a better outcome in that case than silently being unsafe in a way that we have no detection for.

deads2k commented:
/lgtm
/hold

holding for other reviewers.

lavalamp commented:
squashed/rebased

liggitt commented:
/approve
for vendor/modules.txt change

lavalamp commented:
/retest

lavalamp commented:
/retest

lavalamp commented:
This is a concerning amount of flakes, but they don't appear related to this PR.

/retest

deads2k commented:
/lgtm

deads2k commented:
/hold cancel
